pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package featurestates
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/internalapis"
+	featurestatesv1alpha1 "sigs.k8s.io/vsphere-csi-driver/pkg/internalapis/featurestates/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+const (
+	// CRDName represent the name of cnscsisvfeaturestate CRD
+	CRDName = "cnscsisvfeaturestates.cns.vmware.com"
+	// CRDGroupName represent the group of cnscsisvfeaturestate CRD
+	CRDGroupName = "cns.vmware.com"
+	// CRDSingular represent the singular name of cnscsisvfeaturestate CRD
+	CRDSingular = "cnscsisvfeaturestate"
+	// CRDPlural represent the plural name of cnscsisvfeaturestates CRD
+	CRDPlural = "cnscsisvfeaturestates"
+	// WorkLoadNamespaceLabelKey is the label key found on the workload namespace in the supervisor k8s cluster
+	WorkLoadNamespaceLabelKey = "vSphereClusterID"
+	// SVFeatureStateCRName to be used for CR in workload namespaces
+	SVFeatureStateCRName = "svfeaturestates"
+	// crUpdateRetryInterval is the interval at which pending CR update/create tasks are executed
+	crUpdateRetryInterval = 30 * time.Second
+)
+
+// pendingCRUpdates holds latest states of the namespaces to push CR update, latest feature states and the lock to operate
+// on namespaceUpdateMap and latestFeatureStates
+type pendingCRUpdates struct {
+	// lock to acquire before updating namespaceUpdateMap and latestFeatureStates
+	lock *sync.RWMutex
+	// holds namespace name and status to update CR in the namespace
+	// map value true, means update is pending, false means no update is required.
```

Comment:
Why do we need to interpret true or false? If namespace update is not needed why not just remove it from the map?

Reply:
When an update or create is required for workload namespace, I am updating `namespaceUpdateMap map[string]bool` and setting true for that namespace in the map. After update/create is done, I am setting false for that namespace in the map, if failure is observed, the value remains true, so `processPendingCRUpdates`, can process the update in the next retry.

When service is starting for the first time, I am populating `namespaceUpdateMap map[string]bool` with all workload namespaces, with value true, so `processPendingCRUpdates`, can create/update CR in those namespaces.

When a new workload namespace is observed by an informer, I am adding a new namespace in this map with the value set to true, so `processPendingCRUpdates` can help create CR in the new namespace.

when a workload namespace is deleted, I am deleting an entry from `namespaceUpdateMap map[string]bool` , so processPendingCRUpdates does not need to deal pending failed updates for deleted namespace.

maintaining all current workload namespaces in the in-memory map `namespaceUpdateMap`, does not require us to query kubernetes for all workload namespaces when FSS update is observed, we can rely on namespaceUpdateMap, since it is maintained by informer events.

Reply:
Basically, we are not treating the problem as a producer-consumer problem. We are keeping an in-memory cache of all the current namespaces and a state whether it requires FSS replication or not. The downside of approach is more memory usage and processing all the entries in the map every time to see if any namespace needs replication.

The alternate solution could be to treat this as a producer-consume problem where the map has entries only when the FSS is updated or namespace is added. The former is very very rare and only in that case you need to query all SV namespaces.

I would prefer to optimize the solution by using less memory and less processing. But I will leave it to you as I dont want this to be a blocker.

Reply:
> The downside of approach is more memory usage and processing all the entries in the map every time to see if any namespace needs replication.

But this comes with the advantage of reducing additional calls to the API server when config-map changes or workload namespace is added. The memory footprint is very less, as we are just storing fss config-map and workload namespaces name, and their status.